chain/gap/find.go

```diff
@@ -0,0 +1,264 @@
+		})
+	}
+
+	return g.DB.PersistBatch(ctx, skipGaps, heightGaps, taskGaps, nullRounds)
+}
+
+type GapIndexerLens interface {
+	ChainGetTipSetByHeight(ctx context.Context, epoch abi.ChainEpoch, tsk types.TipSetKey) (*types.TipSet, error)
+}
+
+func (g *GapIndexer) findEpochSkips(ctx context.Context, max, min uint64) (visor.GapReportList, error) {
+	log.Debug("finding skipped epochs")
+	reportTime := time.Now()
+
+	var skippedReports []visor.ProcessingReport
+	if err := g.DB.AsORM().ModelContext(ctx, &skippedReports).
+		Order("height desc").
+		Where("status = ?", visor.ProcessingStatusSkip).
+		Where("height >= ?", min).
+		Where("height <= ?", max).
+		Select(); err != nil {
+		return nil, xerrors.Errorf("query processing report skips: %w", err)
+	}
+	log.Debugw("executed find skipped epoch query", "count", len(skippedReports))
+
+	gapReport := make([]*visor.GapReport, len(skippedReports))
+	for idx, r := range skippedReports {
+		gapReport[idx] = &visor.GapReport{
+			Height:     r.Height,
+			Task:       r.Task,
+			Status:     "GAP",
+			Reporter:   g.name,
+			ReportedAt: reportTime,
+		}
+	}
+	return gapReport, nil
+}
+
+func (g *GapIndexer) findEpochGapsAndNullRounds(ctx context.Context, node GapIndexerLens, max, min uint64) (visor.GapReportList, []abi.ChainEpoch, error) {
+	log.Debug("finding epoch gaps and null rounds")
+	reportTime := time.Now()
+
+	var nullRounds []abi.ChainEpoch
+	var missingHeights []uint64
+	res, err := g.DB.AsORM().QueryContext(
+		ctx,
+		&missingHeights,
+		`
+		SELECT s.i AS missing_epoch
+		FROM generate_series(?, ?) s(i)
+		WHERE NOT EXISTS (SELECT 1 FROM visor_processing_reports WHERE height = s.i);
+		`,
+		min, max)
+	if err != nil {
+		return nil, nil, err
+	}
+	log.Debugw("executed find epoch gap query", "count", res.RowsReturned())
+
+	gapReport := make([]*visor.GapReport, 0, len(missingHeights))
+	// walk the possible gaps and query lotus to determine if gap was a null round or missed epoch.
+	for _, gap := range missingHeights {
+		gh := abi.ChainEpoch(gap)
+		tsgap, err := node.ChainGetTipSetByHeight(ctx, gh, types.EmptyTSK)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting tipset by height %d: %w", gh, err)
+		}
+		if tsgap.Height() == gh {
+			log.Debugw("found gap", "height", gh)
+			for _, task := range chain.AllTasks {
+				gapReport = append(gapReport, &visor.GapReport{
+					Height:     int64(tsgap.Height()),
+					Task:       task,
+					Status:     "GAP",
+					Reporter:   g.name,
+					ReportedAt: reportTime,
+				})
+			}
+		} else {
+			log.Debugw("found null round", "height", gh)
+			nullRounds = append(nullRounds, gh)
+		}
+	}
+	return gapReport, nullRounds, nil
+}
+
+type TaskHeight struct {
+	Task   string
+	Height uint64
+	Status string
+}
+
+// TODO rather than use the length of `tasks` to determine where gaps are, use the contents to look for
+// gaps in specific task. Forrest' SQL-Foo isn't good enough for this yet.
```

Comment:
I'd love to get this addressed, just ran out of ideas for how to write it. Being able to only report gaps to the gaps_report_table for specific types of tasks will be useful for our monitoring instances since they don't run with everything enabled. Alternatively filtering could be performed below.

Reply:
Could you weave in clauses like `where task in('messsages','blocks')` and pass in the number of tasks to compare with?

Reply:
I think I could do something like that -- tried here recently but broke things. I'm going to punt that to a follow-on once #655 is addressed and I can automate testing for this with vectors.